### PR TITLE
Updating the release process [ci full]

### PR DESCRIPTION
**Pull request by bendk**
There's two parts to this PR: updating the a-s release process and adding taskcluster code to build artifacts for the new process.

## New release process

We don't really need to "cut releases" anymore, just push commits to release branches.

I modified `automation/prepare-release.py` to handle the new process.  It should create PRs like this:
  - [Updating the release branch for a new rversion](https://github.com/mozilla/application-services/pull/5507)
  - [Updating main to start a new major version](https://github.com/mozilla/application-services/pull/5508)

After those PRs are merged and we've built the release, then you need to tag the version with `automation/tag-release.py`

The new release process simplifies a lot of things and I think that means we can just keep a `CHANGELOG.md` and forget about `CHANGES_UNRELEASED.md`

## Taskcluster stuff

I refactored the taskcluster code more-or-less execute the nightly process, but do it for a real release.  It's not working yet, I'm using this PR to test it.

@ahal couple questions that you can hopefully answer while I'm away:
  - I'm hoping to get this code to run whenever there's a push to a `release-v*` branch.  However, when I tried to test that I didn't see any taskcluster actions.  Do we need to do something to git taskcluster to run on github pushes?  It's also very possible that this is already working and I just did something wrong when testing.
  - I was hoping to use a `kind-dependency` to get the [release-publish](https://github.com/mozilla/application-services/blob/599f7ae8bd7f8b69052239dabe14152903ab8215/taskcluster/ci/release-publish/kind.yml) task to only run after [build-summary](https://github.com/mozilla/application-services/blob/599f7ae8bd7f8b69052239dabe14152903ab8215/taskcluster/ci/build-summary/kind.yml) succeeds.  My thinking is that build-summary only succeeds if all the build task succeed, which means we would only publish the `release.json` file after we knew everything worked.  However,  the `kind-dependency` doesn't seem to translate into an actual dependency in the taskcluster tasks.  `build-publish` only waits on a few tasks then runs.
  
### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: docs/howtos/releases.md
Comment (bendk) [inline docs/howtos/releases.md:23]: I just made up this 2 workdays number, not sure when we should start it.  We should make sure we have plenty of time to finish before firefox-android or firefox-ios want to start cutting their releases.

File: docs/howtos/releases.md
Comment (bendk) [inline docs/howtos/releases.md:41]: Eventually, maybe this could be rolled into the `promote` action.

File: docs/howtos/releases.md
Comment (bendk) [inline docs/howtos/releases.md:47]: And eventually, maybe this could be rolled into the `ship` action.

**Comment by ahal** [comment]
> I'm hoping to get this code to run whenever there's a push to a release-v* branch. However, when I tried to test that I didn't see any taskcluster actions. Do we need to do something to git taskcluster to run on github pushes? It's also very possible that this is already working and I just did something wrong when testing.

This is because we're ignoring pushes to all non-main branches here:
https://github.com/mozilla/application-services/blob/main/.taskcluster.yml#L100

Luckily VPN is set up the way you want, so it should just be a matter of copying them:
https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/.taskcluster.yml#L98

> I was hoping to use a kind-dependency to get the [release-publish](https://github.com/mozilla/application-services/blob/599f7ae8bd7f8b69052239dabe14152903ab8215/taskcluster/ci/release-publish/kind.yml) task to only run after [build-summary](https://github.com/mozilla/application-services/blob/599f7ae8bd7f8b69052239dabe14152903ab8215/taskcluster/ci/build-summary/kind.yml) succeeds. My thinking is that build-summary only succeeds if all the build task succeed, which means we would only publish the release.json file after we knew everything worked. However, the kind-dependency doesn't seem to translate into an actual dependency in the taskcluster tasks. build-publish only waits on a few tasks then runs.

Yeah, all `kind-dependency` does is make sure those kinds are loaded before the current kind. You still need to make the dependency link to whatever specific tasks are desired within the kind dependencies. You can do this by specifying the `dependencies` key manually, using a transform, or in this case, I'd recommend using the [multi_dep](https://github.com/mozilla/application-services/blob/main/taskcluster/app_services_taskgraph/loader/multi_dep.py) loader.

The `multi_dep` loader is designed for cases when you want to automatically create tasks from a set of indeterminate dependencies. Here's an example of how it's used:
https://github.com/mozilla/application-services/blob/main/taskcluster/ci/signing/kind.yml

This will create one `signing` task for every `module-build` task. Come to think of it, you may want your summary task to run after `signing` rather than `module-build`.

Aside: I'm actually working on implementing this functionality in a nicer way directly into core Taskgraph over in [this PR](https://github.com/taskcluster/taskgraph/pull/236), but it's not worth blocking on that.

**Comment by ahal** [comment]
Oh, and if we start running tasks on `releases/*` don't forget to add Github branch protection rules at least as strict as the ones on `main` :).

File: docs/howtos/releases.md
Comment (ahal) [inline docs/howtos/releases.md:19]: Here's [a link](https://firefox-source-docs.mozilla.org/taskcluster/release-promotion.html) to the Gecko docs on the topic if you want to draw inspiration.

File: docs/howtos/releases.md
Comment (ahal) [inline docs/howtos/releases.md:41]: typo: s/haeder/header

File: docs/howtos/releases.md
Comment (ahal) [inline docs/howtos/releases.md:45]: Ideally this will be done via shipit, but until then we may want only Releng to have permission to do this. This way it acts as a poor man's dual signoff.

File: docs/howtos/releases.md
Comment (ahal) [reply docs/howtos/releases.md:47]: Yes, we have a `treescript` which can not only tag releases, but land changes to e.g, increment a version number (if desired).

File: taskcluster/app_services_taskgraph/__init__.py
Comment (ahal) [inline taskcluster/app_services_taskgraph/__init__.py:36]: Gecko calls this parameter `release-type`, might be worth making it consistent.

**Comment by bendk** [comment]
@ahal Thanks for the info on those issues, I just pushed a [commit](https://github.com/mozilla/application-services/pull/5510/commits/ac3444bfc87ab01919998d55ed92dbdd4e54421c) that I think should fix them.

For the dependencies, I went with a transform instead of a loader, since we already implemented this code for `build-summary`.  I think it should work and we can clean it up once that PR you mentioned lands.

I'm going to update the branch protection settings now.

If you think this looks good, then I think we can merge this to main.

File: taskcluster/app_services_taskgraph/target_tasks.py
Comment (ahal) [inline taskcluster/app_services_taskgraph/target_tasks.py:15]: What's the intent of this target task phase exactly? Shouldn't it be called `full` or something?

Fyi I'll be adding new methods for the `release_promotion` aspects, so I'm a bit unclear if this one will still be needed or not.

Comment (bendk) [reply taskcluster/app_services_taskgraph/target_tasks.py:15]: The intention is to run the tasks needed for a release (nightly or otherwise).  `full` is probably a better name, but we also have `pr-full` which is run when you add `[ci: full]` in the PR title.  It's all a bit confusing to me.

Maybe `release` and `pr-full` should both be renamed to `full` and the only difference between a release, nightly, and a PR with `[ci: full]` is the indexing.

File: taskcluster/app_services_taskgraph/__init__.py
Comment (bendk) [reply taskcluster/app_services_taskgraph/__init__.py]: Sounds good, I updated the param.

File: taskcluster/app_services_taskgraph/target_tasks.py
Comment (ahal) [reply taskcluster/app_services_taskgraph/target_tasks.py:15]: Yeah that latter suggestion sounds sensible, though we don't need to worry about it now. I was mostly curious.

For release promotion, we'll schedule only the tasks that are needed for the actual release. Things like builds and tests presumably ran on the push already (e.g, we "promote" an existing build). So fyi this target_tasks method probably won't end up being used there.

File: taskcluster/app_services_taskgraph/transforms/release_publish.py
Comment (ahal) [inline taskcluster/app_services_taskgraph/transforms/release_publish.py:25]: suggestion: you could reduce this to:
```
instance = "production" if config.params["level"] == "3" else "staging"
nightly = "-nightly" if config.params.get("preview-build") else ""
maven_channel = f"maven{nightly}-{instance}"
```

File: taskcluster/app_services_taskgraph/transforms/release_publish.py
Comment (ahal) [inline taskcluster/app_services_taskgraph/transforms/release_publish.py:61]: You *might* want to use the `if-dependencies` feature here:
https://taskcluster-taskgraph.readthedocs.io/en/latest/concepts/task-graphs.html#if-dependencies

In practice soft-dependencies will probably work exactly the same, but the difference is that this task will still run even if all its dependencies get optimized away. But the odds of that happening are very small, so I wouldn't worry about it too much.

**Review by ahal — APPROVED** [review]
Thanks, this looks great! I think we'll need to make a few tweaks for release promotion, but this is a good base to build off of and we can do those in the relpro PR.

File: taskcluster/app_services_taskgraph/target_tasks.py
Comment (bendk) [reply taskcluster/app_services_taskgraph/target_tasks.py]: The target tasks method should be used for the push though right?  I renamed the key to `full` and added a comment on how I think it should work.  Hopefully that's enough good for now.

**Comment by bendk** [comment]
Pushed what's hopefully the last commit.  I tried to clean up / document the target tasks method some.  I also made it so `swift-build` and `nimbus-fml-build` run when `[ci full]` is in the title, but they only use the index when we're doing an actual release build.

Does the last commit seem okay?  I agree that this probably will need to change once we implement release-promotion, but hopefully it's good enough as a base to get started.

File: taskcluster/ci/swift/kind.yml
Comment (ahal) [inline taskcluster/ci/swift/kind.yml:23]: This is the same route from before, so no need to block this PR. Though I notice that there's no `release-type` in the route. This means if you have both staging and prod release types, the former could overwrite the latter. Maybe it's not a problem in this case, but would be good to double check this.

**Review by ahal — APPROVED** [review]
Thanks, this looks good!

File: taskcluster/ci/swift/kind.yml
Comment (bendk) [reply taskcluster/ci/swift/kind.yml:23]: Good point.  I don't think it'll be an immediate issue, but it's good to fix.  I made this issue for it: https://github.com/mozilla/application-services/issues/5545.

I think we can do this as a follow-up, hopefully next week.

**Pull request merged**
